clang/lib/Basic/Offloading.cpp

```diff
@@ -0,0 +1,131 @@
+#include "clang/Basic/Offloading.h"
+
+#include "llvm/ADT/StringRef.h"
+
+#include <algorithm>
+
+namespace clang {
+
+namespace {
+struct OffloadArchToStringMap {
+  OffloadArch arch;
+  const char *arch_name;
+  const char *virtual_arch_name;
+};
+} // namespace
+
+#define SM2(sm, ca) {OffloadArch::SM_##sm, "sm_" #sm, ca}
+#define SM(sm) SM2(sm, "compute_" #sm)
+#define GFX(gpu) {OffloadArch::GFX##gpu, "gfx" #gpu, "compute_amdgcn"}
+static const OffloadArchToStringMap arch_names[] = {
+    // clang-format off
```

Comment:
What does this comment mean?

Reply:
This is upstream code that has been moved here.

Thanks